pkg/backend/openshiftcluster/delete.go

```diff
@@ -24,41 +24,41 @@ func (m *Manager) Delete(ctx context.Context) error {
 		return err
 	}
 
+	m.log.Print("looking for network security groups to remove from subnets")
+	nsgs, err := m.securityGroups.List(ctx, resourceGroup)
+	if err != nil {
+		return err
+	}
+
 	// TODO: ideally we would do this after all the VMs have been deleted
-	for _, subnetID := range []string{
-		m.doc.OpenShiftCluster.Properties.MasterProfile.SubnetID,
-		m.doc.OpenShiftCluster.Properties.WorkerProfiles[0].SubnetID,
-	} {
-		// TODO: there is probably an undesirable race condition here - check if etags can help.
-		s, err := m.subnet.Get(ctx, subnetID)
-		if err != nil {
-			// It is possible that we never properly had access to the subnet,
-			// or that we never joined an NSG to it and it has subsequently been
-			// deleted.  As a result we cannot fail here.  If we do have an NSG
-			// joined to the subnet and an error here prevents us from removing
-			// it, we will fail later in the RG deletion; in principal the user
-			// can fix up and retry.
-			m.log.Print(err)
+	for _, nsg := range nsgs {
+		if nsg.SecurityGroupPropertiesFormat == nil ||
+			nsg.SecurityGroupPropertiesFormat.Subnets == nil {
 			continue
 		}
 
-		nsgID, err := subnet.NetworkSecurityGroupID(m.doc.OpenShiftCluster, subnetID)
-		if err != nil {
-			return err
-		}
+		for _, subnet := range *nsg.SecurityGroupPropertiesFormat.Subnets {
+			// Note: subnet only has value in the ID field,
+			// so we have to make another API request to get full subnet struct
+			// TODO: there is probably an undesirable race condition here - check if etags can help.
+			s, err := m.subnet.Get(ctx, *subnet.ID)
+			if err != nil {
+				return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidLinkedVNet, "", "Failed to get subnet '%s'. Original error: %s.", *subnet.ID, err.Error())
```

Comment:
please can you follow the pattern in deployARMTemplate() to return the original error like this:
```
		b, _ := json.Marshal(err)

		return &api.CloudError{
			StatusCode: http.StatusBadRequest,
			CloudErrorBody: &api.CloudErrorBody{
				Code:    xxx,
				Message: "Xxx.",
				Details: []api.CloudErrorBody{
					{
						Message: string(b),
					},
				},
			},
		}
```